godot-ffi/src/binding/single_threaded.rs

```diff
@@ -0,0 +1,146 @@
+    #[inline(always)]
+    unsafe fn storage() -> &'static Self {
+        static BINDING: BindingStorage = BindingStorage {
+            main_thread: OnceLock::new(),
+            binding: OnceCell::new(),
+        };
+
+        &BINDING
+    }
+
+    /// Initialize the binding storage, this must be called before any other public functions.
+    #[must_use]
+    pub fn initialize(binding: GodotBinding) -> Option<()> {
+        // SAFETY: Either we are the first call to `initialize` and so we are calling from the same thread as ourself. Or we are a later call,
+        // in which case we can tell that the storage has been initialized and dont access `binding`.
+        let storage = unsafe { Self::storage() };
+
+        storage.main_thread.set(std::thread::current().id()).ok()?;
+        storage
+            .binding
+            .set(binding)
+            .ok()
+            .expect("`main_thread` was unset so `binding` should also be unset");
+
+        Some(())
+    }
+
+    /// Get the binding from the binding storage.
+    ///
+    /// # Safety
+    /// - Must be called from the main thread.
+    /// - The binding must be initialized.
+    #[inline(always)]
+    pub unsafe fn get_binding_unchecked() -> &'static GodotBinding {
+        let storage = Self::storage();
+
+        if cfg!(debug_assertions) {
+            let main_thread = storage.main_thread.get().expect(
+                "Godot engine not available; make sure you are not calling it from unit/doc tests",
+            );
+            assert_eq!(main_thread, &std::thread::current().id(), "attempted to access binding from different thread than main thread; this is UB - use the \"experimental-threads\" feature.");
+            storage.binding.get().unwrap()
+        } else {
+            // SAFETY: This function can only be called when the binding is initialized and from the main thread, so we know that it's initialized.
+            unsafe { storage.binding.get().unwrap_unchecked() }
+        }
+    }
+
+    pub fn is_initialized() -> bool {
+        // SAFETY: We do not access `binding`.
+        let storage = unsafe { Self::storage() };
+        storage.main_thread.get().is_some()
+    }
+}
+
+// SAFETY: We ensure that `binding` is only ever accessed from the same thread that initialized it.
+unsafe impl Sync for BindingStorage {}
+// SAFETY: We ensure that `binding` is only ever accessed from the same thread that initialized it.
+unsafe impl Send for BindingStorage {}
+
+pub struct GodotBinding {
+    pub(super) interface: GDExtensionInterface,
+    pub(super) library: GDExtensionClassLibraryPtr,
+    pub(super) global_method_table: BuiltinLifecycleTable,
+    pub(super) class_server_method_table: OnceCell<ClassServersMethodTable>,
+    pub(super) class_scene_method_table: OnceCell<ClassSceneMethodTable>,
+    pub(super) class_editor_method_table: OnceCell<ClassEditorMethodTable>,
+    pub(super) builtin_method_table: OnceCell<BuiltinMethodTable>,
```

Comment:
We have already `OnceCell<GodotBinding>`, why do these need to be in another `OnceCell`? Godot guarantees that the init-levels are invoked exactly once -- the extra cell is a waste of performance.

We could use `Option::unwrap_unchecked()` instead (just like before), which has a validation in Debug mode but is fast in Release.

If you additionally want to ensure it's only set once, a simple `debug_assert!(option.is_none(), "...")` does the job too.

Reply:
There isn't really a difference between doing `OnceCell<T>` and a `Option<T>` that's only written to once, in fact internally a `OnceCell<T>` is just a `UnsafeCell<Option<T>>`. So doing `cell.get().unwrap_unchecked()` and `option.unwrap_unchecked()` should be basically the same thing. In fact writing a quick test for this in [godbolt](https://godbolt.org/z/xEWo568av). I find that the generated assembly is exactly the same for accessing both. The rust code is also not very different, `cell.get().unwrap_unchecked()` vs `option.as_ref().unwrap_unchecked()`. 

So since we're only supposed to be writing to this once, i figured that using a `OnceCell<T>` better communicates intent here. Since `Option<T>` usually doesn't mean "can/will only be written to once". If i saw an `Option` be used in a case like this, especially considering that `OnceCell` exists, then i would likely assume it's meant to be something that can be set and unset over the course of the program, or maybe even never set at all.

Reply:
That argument makes a lot of sense. I missed that you're accessing it with `unwrap_unchecked()` too.

Thanks especially for the Godbolt comparison. To make sure this knowledge isn't lost and we're wondering about optimization in the future, could you write a comment above these declarations along the lines "unchecked OnceCell access is not slower than unchecked Option access and expresses intent better; see <Godbolt link>"?